internal/runtime/catalog/catalog.go

```diff
@@ -19,285 +19,315 @@ package catalog
 import (
 	"fmt"
 	"reflect"
+	goruntime "runtime"
+	"strings"
 
 	"github.com/pkg/errors"
-	"k8s.io/apimachinery/pkg/conversion"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/naming"
 	"k8s.io/kube-openapi/pkg/common"
 )
 
-// Catalog defines methods for working with rpc services and their request and response types.
+// Catalog contains mappings between Go types and the corresponding
+// GroupVersionHooks, metadata and OpenAPIDefinitions.
 type Catalog struct {
-	// scheme allow access api-machinery utilities for handling types and types conversions
+	// scheme allows access to api-machinery utilities to implement conversions of
+	// request and response types.
 	scheme *runtime.Scheme
 
-	// gvhToType and typeToGVH allows tracking relations between hook go types and corresponding gvh.
+	// gvhToType allows mapping a GroupVersionHook to the Go type.
 	gvhToType map[GroupVersionHook]reflect.Type
+
+	// gvhToType allows mapping a Go type to the GroupVersionHook.
 	typeToGVH map[reflect.Type]GroupVersionHook
 
-	// gvhDescriptors allows tracking relations gvh and corresponding descriptors.
-	gvhDescriptors map[GroupVersionHook]hookDescriptor
+	// gvhToHookDescriptor allows mapping a GroupVersionHook to the hook descriptor.
+	gvhToHookDescriptor map[GroupVersionHook]hookDescriptor
+
+	// gvToOpenAPIDefinitions allows mapping a GroupVersion to a func
+	// which returns OpenAPI definitions for all request and response types with that
+	// GroupVersion.
+	gvToOpenAPIDefinitions map[schema.GroupVersion]OpenAPIDefinitionsGetter
 
-	// openAPIDefinitionsGetters holds func returning all the openapi definitions for
-	// request and responses type for each hook defined in a group version.
-	openAPIDefinitionsGetters map[schema.GroupVersion]OpenAPIDefinitionsGetter
+	// catalogName is the name of this catalog. It is set based on the stack of the New caller.
+	// This is useful for error reporting to indicate the origin of the Catalog.
+	catalogName string
 }
 
-// OpenAPIDefinitionsGetter defines the signature of a func returning Open API definitions for
-// request and response types of a given group version.
-// NOTE: the OpenAPIDefinitionsGetter func are automatically generated by openapi-gen.
-type OpenAPIDefinitionsGetter func(ref common.ReferenceCallback) map[string]common.OpenAPIDefinition
+// Hook is a marker interface for a hook.
```

Comment:
nit
```suggestion
// Hook is a marker interface for a hook; Hooks should be defined as a func (*Request runtime.Object, *Response runtime.Object)
```

Reply:
Adde slightly differently the funcs have to use the concrete types not runtime.Object (although those types have to implement those interfaces)